transforms/update-react-imports.js

```diff
@@ -54,16 +80,236 @@ module.exports = function(file, api, options) {
       })
       .size() > 0;
 
-  const reactPath = reactImportPath.paths()[0];
-  const specifiers = reactPath.value.specifiers;
+  // local: imported
+  const reactIdentifiers = {};
+  const reactTypeIdentifiers = {};
+  let canDestructureReactVariable;
+  if (reactImportInUse && isDefaultImport) {
+    // Checks to see if the react variable is used itself (rather than used to access its properties)
+    canDestructureReactVariable =
+      root
+        .find(j.Identifier, {
+          name: 'React',
+        })
+        .filter(path => {
+          return path.parent.parent.value.type !== 'ImportDeclaration';
+        })
+        .filter(
+          path =>
+            !(
+              path.parent.value.type === 'MemberExpression' &&
+              path.parent.value.object.name === 'React'
+            ) &&
+            !(
+              path.parent.value.type === 'QualifiedTypeIdentifier' &&
+              path.parent.value.qualification.name === 'React'
+            ) &&
+            !(
+              path.parent.value.type === 'JSXMemberExpression' &&
+              path.parent.value.object.name === 'React'
+            ),
+        )
+        .size() === 0;
 
-  for (let i = 0; i < specifiers.length; i++) {
-    const specifier = specifiers[i];
-    if (specifier.local.name === 'React') {
-      if (specifier.type === 'ImportNamespaceSpecifier' && !keepReactImport) {
-        j(reactPath).remove();
+    if (canDestructureReactVariable) {
+      // If a variable that we are trying to import has already been declared,
+      // do not try to destructure imports
```

Comment:
What happens in the future when only the destructured imports work? Feels like we should at least drop a warning here that they need to fix it by hand or whatever. _Or_ we should re-alias the import and use a new named version instead.

Reply:
Named imports (both destructured and nondestructured) will always work. We are only planning on deprecating default imports.